### ISSUE TITLE
Jackson serialization is not working properly with polymorphic types stored in a collection

### ISSUE THREAD

**Issue by alphaho** [http4k/http4k#313]
Hi, 

I've found that the current `ConfigurableJackson.autoBody` implementation would not work with collections of polymorphic types. It would lose the `@class` field in the serialized json in the below example:

```kotlin
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "@class")
@JsonSubTypes(
    JsonSubTypes.Type(value = FirstChild::class, name = "first"),
    JsonSubTypes.Type(value = SecondChild::class, name = "second")
)
sealed class PolymorphicParent
data class FirstChild(val something: String): PolymorphicParent()
data class SecondChild(val somethingElse: String): PolymorphicParent()

class SampleTest {
    @Test
    fun `roundtrip polymorphic object to and from body`() {
        val body = Body.auto<PolymorphicParent>().toLens()

        val firstChild: PolymorphicParent = FirstChild("hello")
        val secondChild: PolymorphicParent = SecondChild("world")

        assertThat(body(Response(OK).with(body of firstChild)), equalTo(firstChild))
        assertThat(body(Response(OK).with(body of secondChild)), equalTo(secondChild))
    }

    @Test
    fun `roundtrip list of polymorphic objects to and from body`() {
        val body = Body.auto<List<PolymorphicParent>>().toLens()

        val list = listOf(FirstChild("hello"), SecondChild("world"))
        
        assertThat(body(Response(OK).with(body of list)), equalTo(list))
    }
}
```

While the first test(one without being stored in a collection) works, the second test would get:

```
Caused by: com.fasterxml.jackson.databind.exc.InvalidTypeIdException: Missing type id when trying to resolve subtype of [simple type, class org.http4k.format.PolymorphicParent]: missing type id property '@class'
 at [Source: (String)"[{"something":"hello"},{"somethingElse":"world"}]"; line: 1, column: 22] (through reference chain: java.util.ArrayList[0])
```

**Comment by daviddenton** [comment]
Releasing in 3.199.0 😄

**Comment by daviddenton** [comment]
A quick word of warning before upgrading - using 3.199.0 has caused some breaks in some of my IRL projects. Am investigating what is going on, but I'm seeing these when ```java.lang.reflect.GenericSignatureFormatError: Signature Parse error``` caused by ```	at com.fasterxml.jackson.core.type.TypeReference.<init>(TypeReference.java:33)```

**Comment by daviddenton** [comment]
Unfortunately this has caused a breaking change so I'm going to roll back the change for the polymorphic list and release again.

**Comment by alphaho** [comment]
Is it possible to share an example to reproduce the issue? That may make it easier to understand the issue.

**Comment by daviddenton** [comment]
Sure. Here's a test which shows the difference between the 2 implementations:

```kotlin

    @Test
    fun `write interface implementation to body`() {
        assertThat(Response(OK).with(
            Body.auto<Interface>().toLens() of InterfaceImpl()
        ).bodyString(), equalTo("""{"value":"hello","subValue":"123"}"""))

        assertThat(Response(OK).with(
            Jackson.autoBody<Interface>(writeFn = { writerFor(jacksonTypeRef<Interface>()).writeValueAsString(it) }).toLens() of InterfaceImpl()
        ).bodyString(), equalTo("""{"value":"hello"}"""))
    }
```

**Comment by daviddenton** [comment]
Here's a failing branch with the 3.199.0 code: https://github.com/http4k/http4k/tree/refix-writer-lookup-for-jackson

**Comment by daviddenton** [comment]
I've added an `Body.autoGeneric()` method to cover the polymorphic list case to this branch. Thoughts on if this is a reasonable solution?

**Comment by daviddenton** [comment]
This is now being released in 3.206.0. Thanks for the contribution of the fix @alphaho  :)

**Comment by alphaho** [comment]
Great! Thanks for your help on the PR as well!   =D
